### DISCUSSION TITLE
Promise.all returning whole promise instead of the response

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#325] [Capstones!] [answered]
Gabby 

No other issue tickets. 

Issue: I've got a promise.all which create a new order, and orderItems. Inside the promise.all, I call my createOrder promise and my createOrderItems promise and then I destructure the response so I can pass the order response to setOrders and pass the orderitems response to setOrderItems when I call the promise.all in my form. My promise looks like this: 
```
const createOrderAndOrderItems = (orderTransactionID, items, orderObj) => new Promise((resolve, reject) => {
  const order = createOrder(orderObj);
  const orderItems = items.map((item) => createOrderItem(item, orderTransactionID));

  Promise.all([order, orderItems])
    .then(([orderResp, orderItemsResp]) => {
      resolve({ order: orderResp, orderItems: orderItemsResp });
    }).catch((error) => reject(error));
});
```
All the individual promises are working correctly. They are creating a new order and the orderItems in firebase. The only thing that is not working is, the orderItems response returns the whole createOrderItems promise, not the orderItems array itself. So I cannot pass the response to setOrderItems. 

I also noticed this error when I call my promise.all: 
```
index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
```

What I tried: I console.warned the order response and the orderItems response, and the order response is the orders array from firebase (as expected). The orderItems response returns the whole createOrderItems promise. (the promise is successful, it's just not the response I am expecting). I checked my createOrderItems promise, and in the .then, I am calling getOrderItems. It should be returning the orderItems array.  I also tried doing this: 
```
 createOrderAndOrderItems(order.transactionID, itemInputs, order).then(() => {
      getOrders().then((response) => setOrders(response));
      getOrderItems().then((res) => setOrderItems(res));
    });
```
But I feel like that's overkill. I shouldn't have to call getOrders and getOrderItems again since I am calling those in their respective create promises. It sort of works, but I am getting errors in the 'issues' in my console. ( they don't print the errors though). I think this is something small to resolve. after this, I will be at MVP!

Google links: 
https://stackoverflow.com/questions/47614562/async-await-returning-promise-instead-of-data
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all
https://stackoverflow.com/questions/51338277/async-function-returning-promise-instead-of-value

Link to code: 
https://github.com/Gabrielle-Tobermann/Overture-1/blob/financial-reports/src/components/OrderForm.js
https://github.com/Gabrielle-Tobermann/Overture-1/blob/financial-reports/src/helpers/data/ordersData.js

**Comment by drteresavasquez** [comment]
@Gabrielle-Tobermann try instantiating the function in the promise calls. ` Promise.all([order(), orderItems()])...` OR just don't set it equal to a variable and just run it in the Promise.all...well for that second one, may not work that way. We can take a look after the other tickets.

**Reply by Gabrielle-Tobermann** [reply]
I tried it, but the same thing is happening. The order promise returns the orders array, and the orderItems array returns the createOrderItems promise instead of the orderItems array.

**Reply by drteresavasquez** [reply]
Just jumped off and `orderItems` is returning an array of promises, so you can do a Promise.all for that. Looks like Trin is helping you so you can get that all together! This is a GREAT ticket! Excellent job!

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Solution: I needed a Promise.all for orderItems. (orderItems itself was returning an array of promises) Then, I need to add a .then and retrieve the data as an array, to avoid a nested array. Final result: 
```
const createOrderAndOrderItems = (orderTransactionID, items, orderObj) => new Promise((resolve, reject) => {
  const order = createOrder(orderObj);
  const orderItems = items.map((item) => createOrderItem(item, orderTransactionID));

  const orderItemsPromise = Promise.all(orderItems).then(([resp]) => resp);
  Promise.all([order, orderItemsPromise])
    .then(([orderResp, orderItemsResp]) => {
      resolve({ order: orderResp, orderItems: orderItemsResp });
    }).catch((error) => reject(error));
});
```

Thank you Trinity!
